Bug 1395254 - Consume Taskcluster jobs from standard queue exchanges

armenzg opened the pull request:
Currently, Treeherder consumes Pulse messages from an intermediary service called `taskcluster-treeherder`.
Such service needs to be shut down and its functionality imported into Treeherder.

In order to do this we need to switch to the standard Taskcluster exchanges as defined in here:
https://docs.taskcluster.net/docs/reference/platform/queue/exchanges

On a first pass we are only including the code from `taskcluster-treeherder` without changing
much of Treeherder's code. We will still have an intermediary data structure representing
what `taskcluster-treeherder` was emitting, however, we will stop consuming data directly
from it and be able to shut it down.

Following this work would be to get rid of the intermediary job representation which will
clean up a bunch of code and some old tests.

armenzg commented:
I found the bug making me sad the last two days.
The PR is ready to be reviewed.
Instructions are found in the commit message of the first commit.

I'm now going to focus on updating the tests.

<img width="1610" alt="image" src="https://user-images.githubusercontent.com/44410/59381931-946fab80-8d2a-11e9-942a-7ac0263cf9e6.png">

armenzg commented:
Hi @djmitche , I'm interested on getting your feedback with regards to the usage of the taskcluster libraries. I tried to keep all transferred Python code very similar to the Js code to help the task of reviewing. Thanks!

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+
+def fetchMetadata(taskId):
+    url = taskcluster_urls.api(root_url, "queue", "v1", "task/{taskId}".format(taskId=taskId))
+    return fetch_json(url)
```
This is a bit of a confusing function name -- it's fetching the task definition.  That happens to contain a `metadata` field but it's probably not what you want (`name`, `description`, etc.).  I'd call this function `fetchTask`.

While this approach of constructing a URL manually and fetching it is fine, you could also use a `taskcluster.Queue` object's `task` method to do the same.  The latter comes with some retry logic that `fetch_json` might not have.

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+# treeherder job information in task.extra.treeherder are accepted
+def handleMessage(message, metadata=None):
+    taskId = message["payload"]["status"]["taskId"]
+    taskMetadata = fetchMetadata(taskId) if not metadata else metadata
```
as above, this is the task definition itself, not its metadata.

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+    taskId = message["payload"]["status"]["taskId"]
+    taskMetadata = fetchMetadata(taskId) if not metadata else metadata
+    try:
+        # XXX: I don't understand why message["routes"] does not contain everything needed
```
I'm not sure how the message is formatted here, but in AMQP there's a "primary" route that is typically available as `message.route`, with additional routes included in the Cc header, something like `message.headers.Cc`.  So you *could* search for the route info there, but since you have the task definition anyway (in order to get access to `task.extra`), the approach of looking at `task.routes` is easier.

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+    logger.debug("Message received for task %s with route %s", taskId, str(message["routes"])[0:100])
+
+    # Validation failures are common and logged, so do nothing more.
+    if not validateMessage(taskMetadata):
```
This function also has a confusing name -- it's called `validateMessage` but you're passing the task definition to it.  In the original JS, it was called `validateTask` and that seems a lot more accurate :)

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+        return None
+
+    # build a mapping from exchange name to status
+    eventMap = {
```
maybe a good candidate for a constant?

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+      "exchange/taskcluster-queue/v1/task-exception": "exception",
+    }
+
+    taskType = eventMap[message["exchange"]]
```
You'll probably want to move the `Unknown exchange: ..` up here, or use `.get(..)` instead of `[..]`.

armenzg replied:
I think I addressed this correctly. Let me know if not.

djmitche replied:
Looks perfect :)

djmitche commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,302 @@
+        "links": [],
+        "summary": task["metadata"]["description"],
+      },
+      # XXX: The schema requires this value, however, I can't figure out from the code where this happens
```
Good question!  I see the value in [Pulse Inspector](https://taskcluster-ui.herokuapp.com/pulse-messages?bindings%5B0%5D%5Bexchange%5D=exchange%2Ftaskcluster-treeherder%2Fv1%2Fjobs&bindings%5B0%5D%5Bpattern%5D=%23).  Poking around with `git grep` I finally found it at https://github.com/taskcluster/taskcluster/blob/7c73b2502e6f6a0836cea8dd27584face8d3494c/services/treeherder/src/exchanges.js#L48-L51

The version field could probably be dropped from the schema now that this message format is both produced and consumed in treeherder.  Or just set it here -- it doesn't hurt anything.

armenzg replied:
I dropped it. Thanks for the find!

djmitche commented on treeherder/etl/tasks/pulse_tasks.py:
```diff
@@ -16,7 +17,14 @@ def store_pulse_jobs(pulse_job, exchange, routing_key):
+    # handleMessage expects messages in this format
+    tc_th_message = handleMessage({
+        "exchange": exchange,
+        "routes": routing_key,
```
this is a bit confusing, too -- the key is plural but its value is singular

armenzg replied:
I removed the need to pass `routes` in general.

camd commented on treeherder/etl/job_loader.py:
```diff
@@ -131,6 +131,7 @@ def transform(self, pulse_job):
             platform_src = pulse_job[v] if v in pulse_job else default_platform
             x["job"][k] = self._get_platform(platform_src)
 
+        # XXX: We should handle this in a follow up
```
Maybe this should be prefaced with ``TODO: `` and a bug number for the follow up?

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,295 @@
+
+
+def resultFromRun(jobRun):
+    if jobRun["state"] == "completed":
```
This seems like it could avoid a lot of these ``elif``s by creating a map for ``state`` and ``reasonResolved``.  You could have a map that contains all of "completed", "success", "fail", "canceled" and "superseded".  Then check if ``reasonResolved in map`` exists.  If not, look for ``state in map``, otherwise return "unknown".

Just a thought.  But if you prefer this, fine by me.  :) 

But it seems like at least extracting ``state`` and ``reasonResolved`` from ``jobRun`` so you're not doing it each time, would be preferable.

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,295 @@
+      "buildSystem": "taskcluster",
+      "owner": task["metadata"]["owner"],
+      "taskId": "{taskId}/{runId}".format(taskId=slugid.decode(taskId), runId=runId),
+      # XXX: In a follow up we will use the right name
```
need bug number

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,295 @@
+    if pushInfo["origin"] == "hg.mozilla.org":
+        job["origin"]["pushLogID"] = int(pushInfo["pushId"])
+    else:
+        job["origin"]["pullRequestID"] = pushInfo["pushId"]
```
should this be an ``int`` as well?

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,295 @@
+    return job
+
+
+def handleTaskFailed(pushInfo, task, message):
```
This looks identical to ``handleTaskCompleted``.  Perhaps we don't need both?

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,295 @@
+    return job
+
+
+def handleTaskException(pushInfo, task, message):
```
This is also super similar to ``handleTaskCompleted``.  Perhaps we could merge these three functions to handle the cases?

armenzg replied:
I tried but it was not as straight forward.

I can handle this on a clean up in https://bugzilla.mozilla.org/show_bug.cgi?id=1560596

camd commented:
FWIW: In the ``celery`` ingestion, I get a lot of these messages:

```
Failed validating 'pattern' in schema['properties']['realTaskId']:
    {'description': 'On a follow PR we will replace taskId with realTaskId '
                    'and have a jobGuid instead of the\n'
                    'current taskId\n',
     'maxLength': 50,
     'minLength': 1,
     'pattern': '^[A-Za-z0-9/+-]+$',
     'title': 'Real task ID',
     'type': 'string'}

On instance['realTaskId']:
    'EUn_RFYeQqexO-BVqgKT9g'
```

Is this expected?

armenzg commented:
Hey @camd I've fixed the tests to actually hit the same path as `store_pulse_jobs` and I caught the issue you reported and fixed it. I also did a couple of dry runs to see if there was anything uncaught and fixed whatever I saw. In the celery output you will see some errors for `store_push_data` which is not related to my code.

The tests are all now passing. The fixtures have all been updated and we test about 100 random Pulse messages.

The instructions to test this change are still documented on the [first commit message](https://github.com/mozilla/treeherder/pull/5042/commits/288fa0cd948a6488f6076ab8422d7c4464f88c89).

If you're satisifed with the changes don't land it without me being around and coordinating with the sheriffs. I will back on Wednesday. Thanks!

camd commented:
Great!  I will take a look at this tomorrow.  Thanks for doing this!  :)

camd commented on treeherder/etl/taskcluster_pulse/parse_route.py:
```diff
@@ -34,6 +34,8 @@ def parseRoute(route):
         pushId = parsedRoute[4]
 
-    pushInfo["pushId"] = pushId if int(pushId) else None
+    # XXX: Can you please evaluate I'm doing this right?
```
Yeah, this is a little tricky to parse out.  But it looks right.  Though I fear using the term ``pushId`` here could be confusing since that means something different in Treeherder.  Could you use a better term here?

camd commented:
The new code changes look good to me, but I haven't had a chance to test it locally yet.  I will do that Monday.  Nice work!

armenzg commented:
@djmitche I will have some inline questions in the code in a moment.

@camd I created a django managment script + a normal Python script. Here are the changes since you last looked at the code: https://github.com/mozilla/treeherder/pull/5042/files/fec8ddd8c7629431b23ec9ee48f788a111ee793b..c2279fd3651cb0eb67a2390af2f6cf2c36a908f6

The tests for these changes are *not* currently passing (since switching over to async code) *but* ingesting a push from mozilla-central and comparing it proved to give the same results.

Would you mind looking at the code while I simply fix the Travis test and clean up the scripts to use argparse correctly?

armenzg commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -127,18 +129,14 @@ def handleMessage(message, taskDefinition=None):
-            handleTaskRerun(parsedRoute, task, message["payload"])
-
+        # XXX: The original code used to send an extra Pulse message for the previous run to mark it as "retry"
         job = handleTaskPending(parsedRoute, task, message["payload"])
```
@djmitche I just want your confirmation that removing this code makes sense to you as well.
`taskcluster-treeherder` in this case emits two Pulse messages: one for the current task + one for the previous *run* which was retried.

The message for the previous run (notice `message.runId-1`) would be ingested by Treeherder and would update the previous run as `retry`:

```python
async def handleTaskRerun(pushInfo, task, message):
    let job = this.buildMessage(pushInfo, task, message.runId-1, message);
    job = buildMessage(pushInfo, task, message["runId"]-1, message)
    job["state"] = "completed"
    job["result"] = "fail"
    job["isRetried"] = True   # <-- This is what makes Treeherder mark a job as `retried`
```

Now, all of these comments refer to this diff specifically (a compare with all related commits):
https://github.com/mozilla/treeherder/pull/5042/files/a23e935467dc7667ce7e76bc75a5c1338b6aa246..566d61bf334a44a6e9f24dd7eba3afd45b1b875e

We can know if an excepted task should be considered as retried by checking the  "reasonResolved":
https://github.com/mozilla/treeherder/pull/5042/files/a23e935467dc7667ce7e76bc75a5c1338b6aa246..566d61bf334a44a6e9f24dd7eba3afd45b1b875e#diff-609378e93352a69d9c31bfe55b0a8d64R254
```python
    # Based on https://github.com/taskcluster/taskcluster/blob/1a224a54c17fe7871c39b56e185268082623760f/services/queue/schemas/v1/task-exception-request.yml
    # and https://github.com/taskcluster/taskcluster/blob/f6ed2f3819e4cbaabc8c422852113803b103058f/services/queue/schemas/v1/task-status.yml#L81
    if jobRun["reasonResolved"] in ("worker-shutdown",
                                    "intermittent-task",
                                    "claim-expired"):
        job["isRetried"] = True
```
A problem with this is that we might have new states that would require a retry.

In the scenario where I ingested *all* tasks at once, my changes above seem to make sense, however, once we reach live ingestion I'm not too sure this will work.

I think I could keep the original logic by swapping send Pulse messages (`await this.publishJobMessage(pushInfo, job, message.status.taskId);`) for adding messages to the celery queue.

The other option would be to make `handleMessage` return a list of jobs that need to be processed (instead of just 1).

If you're indiferent as to how I solve it, let me know and I will figure it out. If you have a strong preference let me know.

djmitche replied:
The old code would mark any run other than the most recent as `isRetried`.  While the reasons you identified ('worker-shutdown', etc.) can be a cause of a new run, it's also possible to get a new run with the `rerunTask` API method, even if the previous run was resolved as successful.  So I think the modification you've made would miss that case.  Also, there's a limit to the number of retries a task will do (usually 5), so a task resolved as `worker-shutdown` might *not* be retried if it that run was the last allowed run.

That said, I don't know if this is worth worrying about.  If "isRetried" indicates that a task was *automatically* retried, then arguably it shouldn't be set for `rerunTask` anyway.  But I think the full solution would be one of the two alternatives you mentioned.

armenzg replied:
I have taken the approach of returning a list of jobs.
I have reverted these changes and we're now back to the same code logic we had earlier.

armenzg replied:
I forgot to mention that I stripped the commits we were discussing.
Return list of jobs to be processed as Pulse messages is accomplished in here:
https://github.com/mozilla/treeherder/pull/5042/commits/a29ec9412188803755be5e22de452986dc9a038a

armenzg commented on treeherder/etl/management/commands/ingest_push_and_tasks.py:
```diff
@@ -0,0 +1,119 @@
+            if tc_th_message:
+                logger.info("Loading into DB:\t%s/%s", taskId, run["runId"])
+                # XXX: This seems our current bottleneck
+                JobLoader().process_job(tc_th_message)
```
When I run this script `docker-compose run backend bash -c "./manage.py ingest_push_and_tasks"` I see hundreds of messages that happen within `handleMessage` and later on I will start seeing messages from `JobLoader().process_job(tc_th_message)`.

I wish that all `process_job` were not run last since they become sequential inserts into the DB and are not running concurrently.

Any suggestions on what to try?

armenzg replied:
I think this is bearable. 6 minutes to ingest all tasks for a push is not the end of the world.

armenzg commented:
@djmitche Now, my only code change that changes the original logic is this:
https://github.com/mozilla/treeherder/pull/5042/commits/8c0bfe094bea19c8f3c6ead68647202b03e2f683

This is needed for when ingesting all tasks in one shot.

For live ingestion, even if we create extra jobs for the previous run, there's logic in Treeherder to prevent changing completed jobs:
https://github.com/mozilla/treeherder/blob/master/treeherder/etl/jobs.py#L66-L69

In fact, I need to ingest the runs from the latest to the oldest to make sure that tasks get inserted as "retry" instead of "exception":
https://github.com/mozilla/treeherder/pull/5042/commits/9d2fe737b1629a19bb3f76258a4d9b5813b68ce3

@camd the tests are now passing. The changes since you last reviewed this change can be seen in here: https://github.com/mozilla/treeherder/pull/5042/files/53f96aa79b66643a8eb7ebb1ad135bca90955239..45b0f5b20f0ad2021cbcd0ccc4e137d71c1b6a37

I will keep improving the scripts.

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,340 @@
+            "Task ID: {taskId} Routes: {routes}".format(taskId=taskId, routes=routes)
+        )
+
+    parsedRoute = parseRoute(matchingRoutes[0])
```
You could just return the result of ``parseRoute`` and save a couple lines.

djmitche commented:
I'm not sure what the distinction is when ingesting "all at once" but I trust that you've thought hard about it from the comments, so I'm happy to give my blessing.

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,340 @@
+      },
+    }
+
+    job["origin"] = {
```
It seems like this can be folded into the original definition of ``job`` above.

camd commented on treeherder/etl/taskcluster_pulse/handler.py:
```diff
@@ -0,0 +1,340 @@
+    labels = treeherderConfig.get("labels", [])
+    if not labels:
+        if not treeherderConfig.get("collection"):
+            labels = ["opt"]
```
This is probably fine, and is consistent with the original code for taskcluster-treeherder, iirc.  But I seem to recall having a conversation with @jmaher at some point where he thought the default fall-back should be something other than opt.  I could be mis-remembering, but might be worth checking in with him.  :)

jmaher replied:
there are a lot of different options, mostly I am concerned about the array of options like ccov and asan which might not be worrisome here.  Could we log when we set this by default so we could handle this edge case more gracefully?

armenzg replied:
Hi @jmaher 
At this point any more changes for this I would like to put it to [bug 1560596](https://bugzilla.mozilla.org/show_bug.cgi?id=1560596)

armenzg commented:
@camd @djmitche this is [the last bit](https://github.com/mozilla/treeherder/pull/5042/commits/1a75b78841c0a205af937dace81487a3ec7843e5) I needed to figure out for this PR.
I'm not confident on my skills of running parallel tasks via celery.

It seems to be working locally.

```diff
"""
This module contains tasks related to pulse job ingestion
"""
+import asyncio
import newrelic.agent

from treeherder.etl.job_loader import JobLoader
@@ -14,13 +15,14 @@ def store_pulse_jobs(pulse_job, exchange, routing_key):
    """
    Fetches the jobs pending from pulse exchanges and loads them.
    """
+    loop = asyncio.get_event_loop()
    newrelic.agent.add_custom_parameter("exchange", exchange)
    newrelic.agent.add_custom_parameter("routing_key", routing_key)
    # handleMessage expects messages in this format
-    runs = handleMessage({
+    runs = loop.run_until_complete(handleMessage({
        "exchange": exchange,
        "payload": pulse_job,
-    })
+    }))
    for run in runs:
        JobLoader().process_job(run)
```

djmitche commented:
It's certainly not my area of expertise, either!
